In [1]:
import pandas as pd
import numpy as np
import os

In [3]:
path = r"C:\Users\toloz\Instacart Basket Analysis 9_11_2024"

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', '4.3_orders_products' , 'products.csv'), index_col = False)

In [7]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_wrangled.csv'), index_col = False)

In [11]:
df_test = pd.DataFrame()

In [13]:
 df_test['mix'] = ['a', 'b', 1, True]

In [15]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [21]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [25]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [27]:
df_prods.shape

(49693, 5)

In [29]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [31]:
df_prods_clean.shape

(49677, 5)

In [33]:
#look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [35]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [39]:
df_prods_clean_no_dups.shape

(49672, 5)

In [41]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Directions

In [64]:
# 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.
# Already done in above

### 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
# Tip: Keep an eye on min and max values!

In [69]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#order_dow (Day of the Week):

The minimum value is 0, and the maximum value is 6, which seems expected since days of the week are usually represented by numbers 0-6 (e.g., Sunday = 0, Saturday = 6). Nothing seems off here.
order_hour_of_day (Order Time):

The minimum value is 0, and the maximum is 23, representing the hours in a 24-hour format. This is also reasonable.
days_since_prior_order:

The minimum value is 0, meaning that some orders were placed on the same day as the previous order. The maximum value is 30, which may be a cap for the number of days, but it could be worth checking if this is expected based on your business logic or dataset.
No significant anomalies are immediately apparent in the min or max values of most columns, but further investigation into the purpose of the Unnamed: 0 column and the capping at 30 days for the days_since_prior_order field could be useful.

## 3. Check for mixed-type data in your df_ords dataframe.

In [78]:
# check for mixed-type
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

no mixed type data in df_ords data frame

### 5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [89]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The variable that has missing values is the 'days_since_prior_order' which has 206209 values missing.

### 6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

I will use the median to fill out those missing values
Mainly because :It minimizes the effect of outliers.

It’s a numeric field where imputing with the median can provide a realistic estimate without drastically altering the dataset's overall distribution.

In [113]:
 df_ords['days_since_prior_order'].fillna(7, inplace=True)

In [117]:
# no more missing values
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

### Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [122]:
df_dups_ords = df_ords[df_ords.duplicated()]

In [124]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

 There are no duplicates in this set because it returned empty. but, if there were duplicates I would execute '.drop_duplicates()' to creat a new datafreame without duplicates.

# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [145]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))